src/core/app.lua

```diff
@@ -119,66 +122,138 @@ end
 -- new app network by making the changes needed.
 function configure (new_config)
    local actions = compute_config_actions(configuration, new_config)
-   apply_config_actions(actions, new_config)
-   configuration = new_config
+   apply_config_actions(actions)
    counter.add(configs)
 end
 
 -- Return the configuration actions needed to migrate from old config to new.
---
--- Here is an example return value for a case where two apps must
--- start, one must stop, and one is kept as it is:
---   { start = {'newapp1', 'newapp2'},
---     stop  = {'deadapp1'},
---     keep  = {'oldapp1'},
---     restart = {},
---     reconfig = {}
---   }
 function compute_config_actions (old, new)
-   local actions = { start={}, restart={}, reconfig={}, keep={}, stop={} }
+   local actions = {}
+   local function add_action(action, ...)
+      table.insert(actions, { action, { ... } })
+   end
+
+   -- First determine the links that are going away and remove them.
+   for linkspec in pairs(old.links) do
+      if not new.links[linkspec] then
+         local fa, fl, ta, tl = config.parse_link(linkspec)
+         if old.apps[fa] then add_action('unlink_output', fa, fl) end
+         if old.apps[ta] then add_action('unlink_input', ta, tl) end
+         if not new[linkspec] then add_action('free_link', linkspec) end
```

Comment:
AFAIU also we should always unlink the sides -- if we're throwing away the link, then it was indeed linked before, and surely we should also unlink both sides.